code/modules/reagents/chemistry/machinery/chem_dispenser.dm

```diff
@@ -288,73 +287,92 @@
 			if(!usr.canUseTopic(src, !issilicon(usr)))
 				return
-			if(saved_recipes[name] && tgui_alert(usr, "\"[name]\" already exists, do you want to overwrite it?",, list("Yes", "No")) == "No")
-				return
-			if(name && recording_recipe)
-				for(var/reagent in recording_recipe)
-					var/reagent_id = GLOB.name2reagent[translate_legacy_chem_id(reagent)]
-					if(!dispensable_reagents.Find(reagent_id))
-						visible_message(span_warning("[src] buzzes."), span_hear("You hear a faint buzz."))
-						to_chat(usr, "<span class ='danger'>[src] cannot find <b>[reagent]</b>!</span>")
-						playsound(src, 'sound/machines/buzz-two.ogg', 50, TRUE)
+			if(name && recipe)
+				var/list/first_process = splittext(recipe, ";")
+				if(!LAZYLEN(first_process))
+					return
+				var/resmismatch = FALSE
+				for(var/reagents in first_process)
+					var/list/reagent = splittext(reagents, "=")
+					var/amt = text2num(reagent[2]) || 0
+					var/reagent_id = GLOB.name2reagent[translate_legacy_chem_id(reagent[1])]
+					if(dispensable_reagents.Find(reagent_id) || (reagent_id && amt < 0))
+						if (!resmismatch && !check_macro_part(reagents))
+							resmismatch = TRUE
+						continue
+					else
+						var/chemid = reagent[1]
+						visible_message("<span class='warning'>[src] buzzes.</span>", "<span class='italics'>You hear a faint buzz.</span>")
+						to_chat(usr, "<span class='danger'>[src] cannot find Chemical ID: <b>[chemid]</b>!</span>")
+						playsound(src, 'sound/machines/buzz-two.ogg', 50, 1)
 						return
-				saved_recipes[name] = recording_recipe
-				recording_recipe = null
-				. = TRUE
-		if("cancel_recording")
-			if(!is_operational)
-				return
-			recording_recipe = null
-			. = TRUE
+				if (resmismatch && alert("[src] is not yet capable of replicating this recipe with the precision it needs, do you want to save it anyway?",, "Yes","No") == "No")
+					return
+				saved_recipes += list(list("recipe_name" = name, "contents" = recipe))
 		if("reaction_lookup")
 			if(beaker)
 				beaker.reagents.ui_interact(usr)
 
+/obj/machinery/chem_dispenser/proc/check_macro(macro)
+	for (var/reagent in splittext(trim(macro), ";"))
+		if (!check_macro_part(reagent))
+			return FALSE
+	return TRUE
+
+/obj/machinery/chem_dispenser/proc/check_macro_part(part, res = macroresolution)
+	var/detail = splittext(part, "=")
+	return is_resolution(text2num(detail[2]), res)
+
+/obj/machinery/chem_dispenser/proc/is_resolution(amt, res = macroresolution)
+	. = FALSE
+	for(var/i in 5 to macroresolution step -1)
+		if(amt % i == 0)
+			return TRUE
+
+/obj/machinery/chem_dispenser/proc/process_recipe_list(list/recipe)
+	var/list/key_list = list()
+	var/list/final_list = list()
+	var/list/first_process = splittext(recipe["contents"], ";")
+	for(var/reagents in first_process)
+		var/list/splitreagent = splittext(reagents, "=")
+		final_list += list(avoid_assoc_duplicate_keys(splitreagent[1],key_list) = text2num(splitreagent[2]))
```

Comment:
`final_list[avoid_assoc_duplicate_keys(splitreagent[1], key_list)] = text2num(splitreagent[2])`
You REALLY don't know how to use assoc lists.

Reply:
yeah, I already said this

Reply:
etra you coon there's no `recording_recipe` in the whole code anymore